### ISSUE TITLE
Filesystem corruption happening more often than we anticipated

### ISSUE THREAD

**Issue by gideonthomas** [humphd/brackets#450]
See https://news.ycombinator.com/item?id=10151485

**Comment by humphd** [comment]
I think that's what this is: https://github.com/mozilla/thimble.webmaker.org/issues/1031

cc @modeswitch.  We're seeing `EFILESYSTEMERROR` in Firefox in the wild.  None of us can reproduce it.  I wonder if it's older versions not letting them run the db across a third-party iframe?

**Comment by modeswitch** [comment]
From the other issues, it's almost certainly because he's in private browsing mode. Filer needs to be more polite when that happens, by providing a useful error message for the application to pass on to the user, and maybe even by providing a reasonable fallback if possible.

Users on older versions of Firefox will also have this problem, and I don't think there's an easy way of distinguishing between those and the private browsers.

**Comment by humphd** [comment]
Firefox throws an `InvalidStateError` in Private Browsing mode.  So we should probably wrap the call to IndexedDB `open` in a `try...catch` and send a better error up to Filer, `ENOTPERMITTED` or something that you can get in the app and provide UI.

See https://github.com/mozilla/localForage/issues/363#issuecomment-128667598

**Comment by humphd** [comment]
I spent a very frustraging afternoon trying to fix this in Filer, and I don't think it can be cleanly handled.  Firefox throws when you try to `indexeddb.open()`, and won't allow you to catch the exception.  It also means that your `request.onerror` handler never gets called.  So you can't catch the error, and you can't get far enough for an error handler to fire.  It's bonkers that they've chosen to do this.

I've asked for advice in https://bugzilla.mozilla.org/show_bug.cgi?id=781982#c27

**Comment by modeswitch** [comment]
Yeah, that seems broken. I caught myself up on the bug you linked. I'm not sure there's much we can do in Filer until the issue is addressed on the platform side.

**Comment by humphd** [comment]
Josh suggests to put this in an iframe on startup:

```js
var canUseIndexedDB = false;
function notifyParent() {
  window.parent.reportIndexedDBStatus(canUseIndexedDB);
}
setTimeout(notifyParent, 0);
window.indexedDB;
canUseIndexedDB = true;
```
